dialogue-annotations-processor/src/test/java/com/palantir/myservice/service/MyService.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.myservice.service;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.HttpMethod;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.annotations.DialogueClient;
+import com.palantir.dialogue.annotations.Request;
+import java.util.UUID;
+
+@DialogueClient
+public interface MyService {
+
+    @Request(method = HttpMethod.POST, path = "/greet")
+    String greet(
+            // Idea is that Json.class provides the encoder to transform the
+            // greeting string into application/json. Another hand-written
+            // CustomJson.class may provide a custom ObjectMapper, but we
+            // should make it relatively easy to implement this sort of thing.
+            @Request.Body String greeting);
+
+    // Support blocking and listenablefuture based on the return type
+    @Request(method = HttpMethod.GET, path = "/greeting", accept = CustomStringDeserializer.class)
```

Comment:
I don't think we'll want to annotate directly with deserializer classes, rather with some kind of DeserializerFactory. This way we get some type safety while retaining the ability to reuse the jackson ObjectMapper using base-type Object.

This differs slightly from conjure, were we have a factory for all our types which we know are json serializable, in this case we want to safely support custom per-type deserializers.

```java
@Request(method = HttpMethod.GET, path = "/greeting", accept = CustomStringDeserializer.class)
String getGreetingAsync();

interface DeserializerFactory<T> {
    <U extends T> Deserializer<U> deserializerFor(TypeMarker<U> type);
}

class CustomStringDeserializer implements DeserializerFactory<String> {
   Deserializer<String> deserializerFor(TypeMarker<String> type);
}
```

Reply:
@carterkozak I'm not sure I understand how users would implement the `DeserializerFactory<T>`. There is a fair bit of logic in https://github.com/palantir/dialogue/blob/develop/dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java#L234-L265 for handling errors and various status codes and that users would conceivably want to take advantage of but would not necessarily have access to without the current clientFactory's `ConjureRuntime`

Reply:
Yep, we will need to provide some basic building blocks (which likely delegate to our default conjureruntime) to make custom encodings easier, however I don't want to prevent folks from using custom encodings that are capable of supporting non-conjure use-cases like mapping a 404 response to an empty optional.

Reply:
I think that if you passed the clientFactories `ConjureRuntime` into the deserializerFactory then you could relatively cleanly implement what you want. I would stay wasay from using the default conjure runtime since it would prevent us from taking advantage of any additional configuration injected by WC

Reply:
That's a good point. I like the idea of sharing customization and monitoring with the wc clientfactory, but we'll have to be careful not to couple too closely with conjure, otherwise we may inadvertently modify behavior of custom clients without realizing it. For example, the ConjureRuntime doesn't provide any guarantee that data will be encoded to JSON. We want to allow folks to bring their own ObjectMapper for custom type handling which isn't something I'd be thrilled about using with proper conjure dialogue clients.

Reply:
Ok, I think I understand the concern: you'd like to be able to write one deserializer for Object that can handle anything it wants. So we need the TypeMarker so the information can be retained at runtime.

Reply:
I went for DeserializerFactory without the param, as I couldn't write a simple StdSerializer without it, but it might be that I suck at generics.

Reply:
Hmm, that doesn't quite work either. Let me see what I can do

Reply:
I think it's the opposite case where we want the parameter: If I need to send an encoded `Foo` object to a remote server, I don't need to support serializing every type. In that case we'd want an error if I tried to serialize an `Object` or `Bar`, but work properly if I serialize a `Foo`.

Reply:
Yeah, I think the thing that was problematic was the "<U extends T>". I tried it without and it seems to work nicer, but I'm yet to try to implement the Json serializer/deserializer to see if that breaks something.

Reply:
The ``ConjureRuntime`` piece: I don't feel we want to pass that in, because it's got the ability to do actual RPC - so a Serializer/Deserializer should not have access to that? Maybe BodySerDe? But I don't see any customization of that in Witchcraft, so I can instantiate that directly in my Json implementation.